### PR TITLE
clear reader cache in background to prevent leak

### PR DESCRIPTION
**Pull request by dsyme**
@TIHan This eliminates the long-term memory leak from the IL readers used by each TP implementation by clearing the reader cache every 30 seconds since last access (at the potential cost of re-computing readers).

### CODE REVIEW

**Comment by dsyme** [comment]
@TIHan Not sure the best way to unit test this.  Perhaps an internal-only test hook

File: src/ProvidedTypes.fs [hunk 1]
```diff
@@ -6652,12 +6677,36 @@ namespace ProviderImplementation.ProvidedTypes.AssemblyReader
 
         let GetReaderCache () = ReadOnlyDictionary(readerCache)
 
+        // Auto-clear the cache every 30.0 seconds.
+        // We would use System.Runtime.Caching but some version constraints make this difficult.
+        let enableAutoClear = try Environment.GetEnvironmentVariable("FSHARP_TPREADER_AUTOCLEAR_OFF") = null with _ -> true
+        let clearSpanDefault = 30000
```
Comment (cartermp) [inline src/ProvidedTypes.fs]: What's the significance of 30 seconds compared with other values? When we consider that FCS re-checks frequently, how often would this re-allocate the large amount of data (400+ MB) on the LOH in normal usage?

**Comment by TIHan** [comment]
Re testing: Yes, we would need to find an internal hook, similar to how I did testing before with the `#if` directives.

While this is to prevent memory leaks, it may not actually prevent it as each time we fetch assembly DLLs, we read a full `byte []` which can allocate on the LOH. Over time, this can keep expanding the LOH depending on how bad the fragmentation is. Perhaps, we should address that issue in a separate PR.

I agree with @cartermp - it would be nice if we had some sort of interface that would allow us to hook into the lifetime of metadata from outside the TP (such as Roslyn). If that isn't feasible, then this may be the best we got.

**Comment by dsyme** [comment]
> Re testing: Yes, we would need to find an internal hook, similar to how I did testing before with the #if directives.

OK will add this

> While this is to prevent memory leaks, it may not actually prevent it as each time we fetch assembly DLLs, we read a full byte [] which can allocate on the LOH. Over time, this can keep expanding the LOH depending on how bad the fragmentation is. Perhaps, we should address that issue in a separate PR.

Yes, agreed.  Will add separate issue for use of a big `byte[]`

> I agree with @cartermp - it would be nice if we had some sort of interface that would allow us to hook into the lifetime of metadata from outside the TP (such as Roslyn). If that isn't feasible, then this may be the best we got.

Yes indeed, agreed. Perhaps we can discuss as part of https://github.com/fsprojects/FSharp.TypeProviders.SDK/issues/297.  This issue really says "how can the TP know about the target context" - at the moment the only information communicated is the awful reflection hack for the list of referenced assemblies.

**Comment by cartermp** [comment]
@dsyme see here for issue on the `byte[]` array: https://github.com/Microsoft/visualfsharp/issues/5931

**Comment by dsyme** [comment]
@TIHan @cartermp I updated this so that it uses both a weak and strong cache

1. the weak cache ensures sharing while a type provider is active, if it creates the same reader more than once, and also sharing while anything else is keeping the instance alive. The weak cache is not auto-cleared.

2. the strong cache extends the lifetime of the readers by 30-60 seconds to improve sharing as TP instances get recreated over the same base set of DLLs.  It gets auto-cleared 30 seconds after last use.

These fixes are still tactical.

The 30 second timeout is just based on an estimate that that is long enough to re-share TP instances under changes to a project or re-creation of a TP instance.

**Comment by dsyme** [comment]
@TIHan I had trouble in the test checking the weak entries were cleared. THey _seem_ to be cleared when running the tests as a console program, but when running using `dotnet test` I have trouble.

I assume this is something to do with how `dotnet test` runs, or that some other kind of handle is being kept.  But I've checked the ProvidedTypes code again and again and can't see where any global handles are being kept - which jives with the fact that the test passes when run as a command line.

**Pull request merged**